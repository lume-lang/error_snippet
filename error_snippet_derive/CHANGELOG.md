# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.5](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.4...error_snippet_derive-v0.1.5) - 2025-06-14

### Added

- *(derive)* allow merging spans and labels into single field

## [0.1.4](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.3...error_snippet_derive-v0.1.4) - 2025-06-12

### Fixed

- *(derive)* correctly handle lifetimes on error structs

## [0.1.3](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.2...error_snippet_derive-v0.1.3) - 2025-05-27

### Added

- add `#[cause]` attribute

## [0.1.2](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.1...error_snippet_derive-v0.1.2) - 2025-05-11

### Fixed

- fix!(derive): use dynamic errors for #[related]

## [0.1.1](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.0...error_snippet_derive-v0.1.1) - 2025-05-11

### Other

- move repository to `lume-lang`
