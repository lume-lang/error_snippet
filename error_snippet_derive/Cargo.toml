[package]
name = "error_snippet_derive"
version = "0.1.6"
authors = ["Max T. Kristiansen <me@maxtrier.dk>"]
description = "Derive macros for `error_snippet`."
repository = "https://github.com/lume-lang/error_snippet"
license = "MIT"
edition = "2018"
rust-version = "1.85.0"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0.83"
quote = "1.0.35"
syn = "2.0.87"

[dev-dependencies]
insta = "=1.43.1"
