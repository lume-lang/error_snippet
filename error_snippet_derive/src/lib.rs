use diagnostic::Diagnostic;
use quote::quote;
use syn::{parse_macro_input, DeriveInput};

mod args;
mod diagnostic;
mod fields;
mod fmt;
mod tokens;

#[proc_macro_derive(Diagnostic, attributes(diagnostic, span, label, related, cause))]
pub fn derive_diagnostic(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    let cmd = match Diagnostic::from(input) {
        Ok(cmd) => cmd.tokens().unwrap_or_else(|err| err.to_compile_error()),
        Err(err) => return err.to_compile_error().into(),
    };

    quote!(#cmd).into()
}
