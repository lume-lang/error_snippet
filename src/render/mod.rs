use crate::Diagnostic;

pub mod graphical;

pub use graphical::*;

/// Represents a wrapper around a standard formatter.
pub struct Formatter<'a> {
    inner: &'a mut dyn std::fmt::Write,
}

impl std::fmt::Write for Formatter<'_> {
    fn write_str(&mut self, s: &str) -> std::fmt::Result {
        self.inner.write_str(s)
    }
}

/// Defines a trait for rendering diagnostics to a formatter.
pub trait Renderer {
    /// Renders the diagnostic to a string buffer.
    fn render(&mut self, diagnostic: &dyn Diagnostic) -> Result<String, std::fmt::Error> {
        let mut buffer = String::new();
        let mut formatter = Formatter { inner: &mut buffer };

        self.render_fmt(&mut formatter, diagnostic)?;

        Ok(buffer)
    }

    /// Renders the diagnostic to the standard output buffer.
    fn render_stderr(&mut self, diagnostic: &dyn Diagnostic) -> std::fmt::Result {
        eprint!("{}", self.render(diagnostic)?);

        Ok(())
    }

    /// Renders the diagnostic to the given formatter.
    fn render_fmt(&mut self, f: &mut Formatter, diagnostic: &dyn Diagnostic) -> std::fmt::Result;
}
