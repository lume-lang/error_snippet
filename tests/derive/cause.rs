use std::{ops::Range, sync::Arc};

use error_snippet::NamedSource;
use error_snippet_derive::Diagnostic;
use insta::assert_snapshot;

use crate::render;

#[test]
fn single_cause() {
    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "child error")]
    struct Child {}

    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "parent error")]
    struct Parent {
        #[cause]
        pub children: Vec<error_snippet::Error>,
    }

    assert_snapshot!(render(Parent {
        children: vec![Child {}.into()]
    }));
}

#[test]
fn multiple_causes() {
    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "child error")]
    struct Child {}

    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "parent error")]
    struct Parent {
        #[cause]
        pub children: Vec<error_snippet::Error>,
    }

    assert_snapshot!(render(Parent {
        children: vec![Child {}.into(), Child {}.into()]
    }));
}

#[test]
fn cause_with_source() {
    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "child error")]
    struct Child {
        #[span]
        pub source: Arc<NamedSource>,

        #[label("type not allowed")]
        pub span: Range<usize>,
    }

    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "parent error")]
    struct Parent {
        #[cause]
        pub children: Vec<error_snippet::Error>,
    }

    let source = Arc::new(NamedSource::new(
        "some_file.lm",
        r#"fn main() -> void {
    return 0;
}
"#,
    ));

    assert_snapshot!(render(Parent {
        children: vec![Child {
            source,
            span: 13..17
        }
        .into()]
    }));
}

#[test]
fn sourced_error_with_cause() {
    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "child error")]
    struct Child {}

    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "parent error")]
    struct Parent {
        #[cause]
        pub children: Vec<error_snippet::Error>,

        #[span]
        pub source: Arc<NamedSource>,

        #[label("type not allowed")]
        pub span: Range<usize>,
    }

    let source = Arc::new(NamedSource::new(
        "some_file.lm",
        r#"fn main() -> void {
    return 0;
}
"#,
    ));

    assert_snapshot!(render(Parent {
        children: vec![Child {}.into()],
        source,
        span: 13..17
    }));
}

#[test]
fn multiple_causes_with_source() {
    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "child error")]
    struct Child {
        #[span]
        pub source: Arc<NamedSource>,

        #[label("type not allowed")]
        pub span: Range<usize>,
    }

    #[derive(Debug, Diagnostic)]
    #[diagnostic(message = "parent error")]
    struct Parent {
        #[cause]
        pub children: Vec<error_snippet::Error>,
    }

    let source = Arc::new(NamedSource::new(
        "some_file.lm",
        r#"fn main() -> void {
    return 0;
}
"#,
    ));

    assert_snapshot!(render(Parent {
        children: vec![
            Child {
                source: source.clone(),
                span: 13..17
            }
            .into(),
            Child {
                source: source.clone(),
                span: 24..30
            }
            .into()
        ]
    }));
}
