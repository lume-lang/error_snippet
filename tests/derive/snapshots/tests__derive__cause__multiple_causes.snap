---
source: tests/derive/cause.rs
expression: "render(Parent { children: vec![Child {}.into(), Child {}.into()] })"
---
× error: parent error
      × error: child error

      × error: child error
